server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -480,17 +482,38 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId) {
                     // snapshotting indices of the same name
                     logger.debug(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
                         "but failed to clean up its index folder due to the directory not being empty.", metadata.name(), indexId), dnee);
+                    deleteSuccess = false;
                 } catch (IOException ioe) {
                     // a different IOException occurred while trying to delete - will just log the issue for now
                     logger.debug(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
                         "but failed to clean up its index folder.", metadata.name(), indexId), ioe);
+                    deleteSuccess = false;
+                }
+                if (deleteSuccess) {
```

Comment:
Not adding any logic to deal with the failure scenario here because that is a fairly complex problem to automate (basically you'd want to recursively delete an index path once you're sure no more blobs will be added under it, this is trivial to do manually but non-trivial to do automatically).